src/main/java/org/graylog/plugins/usagestatistics/UsageStatsOptOutService.java

```diff
@@ -0,0 +1,153 @@
+import com.squareup.okhttp.Headers;
+import com.squareup.okhttp.HttpUrl;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.RequestBody;
+import com.squareup.okhttp.Response;
+import org.graylog.plugins.usagestatistics.providers.SmileObjectMapper;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.cluster.ClusterId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.ws.rs.core.HttpHeaders;
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import static org.graylog.plugins.usagestatistics.UsageStatsConstants.CONTENT_TYPE;
+import static org.graylog.plugins.usagestatistics.UsageStatsConstants.USAGE_STATS_VERSION;
+import static org.graylog.plugins.usagestatistics.UsageStatsConstants.USER_AGENT;
+
+public class UsageStatsOptOutService {
+    private static final Logger LOG = LoggerFactory.getLogger(UsageStatsOptOutService.class);
+
+    private final ClusterConfigService clusterConfigService;
+    private final UsageStatsConfiguration config;
+    private final Executor executor;
+    private final OkHttpClient httpClient;
+    private final ObjectMapper objectMapper;
+
+    @Inject
+    public UsageStatsOptOutService(ClusterConfigService clusterConfigService,
+                                   UsageStatsConfiguration config,
+                                   OkHttpClient httpClient,
+                                   @SmileObjectMapper ObjectMapper objectMapper) {
+        this(clusterConfigService, config, httpClient, objectMapper, Executors.newSingleThreadExecutor());
+    }
+
+    protected UsageStatsOptOutService(ClusterConfigService clusterConfigService,
+                                      UsageStatsConfiguration config,
+                                      OkHttpClient httpClient,
+                                      ObjectMapper objectMapper,
+                                      Executor executor) {
+        this.clusterConfigService = clusterConfigService;
+        this.config = config;
+        this.executor = executor;
+        this.httpClient = httpClient;
+        this.objectMapper = objectMapper;
+    }
+
+    public UsageStatsOptOutState getOptOutState() {
+        return clusterConfigService.getOrDefault(UsageStatsOptOutState.class, UsageStatsOptOutState.create(false));
+    }
+
+    public void createOptOut(final UsageStatsOptOutState optOutState) {
+        if (optOutState == null) {
+            return;
+        }
+        if (!optOutState.isOptOut()) {
+            LOG.debug("Disable opt-out in cluster config");
+            clusterConfigService.write(optOutState);
+            return;
+        }
+
+        // Run the opt-out request outside of the calling thread so it does not block.
+        executor.execute(new Runnable() {
```

Comment:
Instead of using a "custom" executor and a `Runnable` here we could simply use `httpClient.newCall(…).enqueue(…)`, see https://github.com/square/okhttp/wiki/Recipes#asynchronous-get and [`Callback`](https://square.github.io/okhttp/2.x/okhttp/com/squareup/okhttp/Callback.html).